# Problem 31
 [Source](https://projecteuler.net/problem=31)

In the United Kingdom the currency is made up of pound (£) and pence (p). There are eight coins in general circulation:

> 1p, 2p, 5p, 10p, 20p, 50p, £1 (100p), and £2 (200p).

It is possible to make £2 in the following way:

> 1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p

How many different ways can £2 be made using any number of coins?

In [5]:
# Problem 31 workspace
count = 1
#inelegant but fast enough
for p1 in range(0, 201):
    print("p1: ", p1)
    for p2 in range(0, 101):
        # print(p2)
        for p5 in range(0, 41):
            for p10 in range(0, 21):
                for p20 in range(0, 11):
                    for p50 in range(0, 5):
                        for p100 in range(0, 3):
                            if p1 + 2*p2 + 5*p5 + 10*p10 + 20*p20 + 50*p50 + 100*p100 == 200:
                                count += 1
count

p1:  0
p1:  1
p1:  2
p1:  3
p1:  4
p1:  5
p1:  6
p1:  7
p1:  8
p1:  9
p1:  10
p1:  11
p1:  12
p1:  13
p1:  14
p1:  15
p1:  16
p1:  17
p1:  18
p1:  19
p1:  20
p1:  21
p1:  22
p1:  23
p1:  24
p1:  25
p1:  26
p1:  27
p1:  28
p1:  29
p1:  30
p1:  31
p1:  32
p1:  33
p1:  34
p1:  35
p1:  36
p1:  37
p1:  38
p1:  39
p1:  40
p1:  41
p1:  42
p1:  43
p1:  44
p1:  45
p1:  46
p1:  47
p1:  48
p1:  49
p1:  50
p1:  51
p1:  52
p1:  53
p1:  54
p1:  55
p1:  56
p1:  57
p1:  58
p1:  59
p1:  60
p1:  61
p1:  62
p1:  63
p1:  64
p1:  65
p1:  66
p1:  67
p1:  68
p1:  69
p1:  70
p1:  71
p1:  72
p1:  73
p1:  74
p1:  75
p1:  76
p1:  77
p1:  78
p1:  79
p1:  80
p1:  81
p1:  82
p1:  83
p1:  84
p1:  85
p1:  86
p1:  87
p1:  88
p1:  89
p1:  90
p1:  91
p1:  92
p1:  93
p1:  94
p1:  95
p1:  96
p1:  97
p1:  98
p1:  99
p1:  100
p1:  101
p1:  102
p1:  103
p1:  104
p1:  105
p1:  106
p1:  107
p1:  108
p1:  109
p1:  110
p1:  111
p1:  112
p1:  113
p1:  114
p1:  115
p1:  116
p1:  117
p1:  118
p1:  119
p1:  120
p1:  121
p1:  122
p1:

73682

In [6]:
count

73682

In [2]:
200*100*40*21*11*5*3*2

5544000000

## Answer: 73682

___

# Problem 32
 [Source](https://projecteuler.net/problem=32)

We shall say that an $n$-digit number is pandigital if it makes use of all the digits $1$ to $n$ exactly once; for example, the $5$-digit number, $15234$, is $1$ through $5$ pandigital.

The product $7254$ is unusual, as the identity, $39 \times 186 = 7254$, containing multiplicand, multiplier, and product is $1$ through $9$ pandigital.

Find the sum of all products whose multiplicand/multiplier/product identity can be written as a $1$ through $9$ pandigital.

HINT: Some products can be obtained in more than one way so be sure to only include it once in your sum.

In [ ]:
pan_dig = set(range(1,10))

def get_digits(n_str):
    return set(map(int, n_str))

# get_digits(156)

TypeError: 'int' object is not iterable

In [3]:
1e5

100000.0

In [23]:
# Problem 32 workspace
def get_digits(n: int):
    return set(map(int, str(n)))

count = 0
pairs = []

for i in range(1, int(1e2)):
    # if i%1e3 == 0:
    #     print(i)    
    for j in range(i+1, int(1e4)):
        p = i*j
        num_string = str(p) + str(i) + str(j)
        if len(num_string) == 9 and get_digits(num_string) == pan_dig:
            print("increment")
            count += 1
            pairs.append((i, j))

count
        

increment
increment
increment
increment
increment
increment
increment
increment
increment


9

In [14]:
pairs

[(4, 1738),
 (4, 1963),
 (12, 483),
 (18, 297),
 (27, 198),
 (28, 157),
 (39, 186),
 (42, 138),
 (48, 159)]

In [16]:
4*1738 

6952

In [17]:
sum(set([a*b for (a,b) in pairs]))

45228

## Answer: 45228

___

# Problem 33
 [Source](https://projecteuler.net/problem=33)

The fraction $49/98$ is a curious fraction, as an inexperienced mathematician in attempting to simplify it may incorrectly believe that $49/98 = 4/8$, which is correct, is obtained by cancelling the $9$s.

We shall consider fractions like, $30/50 = 3/5$, to be trivial examples.

There are exactly four non-trivial examples of this type of fraction, less than one in value, and containing two digits in the numerator and denominator.

If the product of these four fractions is given in its lowest common terms, find the value of the denominator.

In [27]:
# Problem 33 workspace
from fractions import Fraction

for i in range(10, 100):
    for j in range(11, 100):
        if i%10==0 and j%10 == 0:
            continue
        if i>=j:
            continue
        f = Fraction(i, j)
        
        a, b = f.as_integer_ratio()
        i_dig = get_digits(i)
        j_dig = get_digits(j)
        if a in get_digits(i) and b in get_digits(j):
            print(i, j)


11 22
11 33
11 44
11 55
11 66
11 77
11 88
11 99
12 24
12 36
12 48
13 26
13 39
13 65
14 28
15 75
16 32
16 64
16 96
17 85
19 95
21 42
21 63
21 84
22 33
22 55
22 77
22 99
24 36
26 39
26 65
31 62
31 93
33 44
33 55
33 77
33 88
36 48
39 65
41 82
42 63
44 55
44 77
44 99
55 66
55 77
55 88
55 99
62 93
63 84
66 77
77 88
77 99
88 99


In [ ]:
#19/95
#16/64
#26/65
#49/98

0.2

In [36]:
Fraction(19,95)*Fraction(16, 64)*Fraction(26,65)*Fraction(49,98)

Fraction(1, 100)

## Answer: 100

___

# Problem 34
 [Source](https://projecteuler.net/problem=34)

$145$ is a curious number, as $1! + 4! + 5! = 1 + 24 + 120 = 145$.

Find the sum of all numbers which are equal to the sum of the factorial of their digits.

Note: As $1! = 1$ and $2! = 2$ are not sums they are not included.

In [144]:
# Problem 34 workspace
def naive_factorial(n):
    if n == 1 or n == 0:
        return 1
    return n*naive_factorial(n-1)

fact_dict = {n: naive_factorial(n) for n in range(10)}

fact_dict



{0: 1, 1: 1, 2: 2, 3: 6, 4: 24, 5: 120, 6: 720, 7: 5040, 8: 40320, 9: 362880}

In [ ]:
[fact_dict[n] for n in get_digits(i)]

[1, 2, 6]

In [155]:
fact_sum = []
for i in range(10, int(1e7)):
    if sum([fact_dict[n] for n in get_digits(i)]) == i:
        fact_sum.append(i)

len(fact_sum)


2

In [156]:
sum(fact_sum)

40730

## Answer: 40730

___

# Problem 35
 [Source](https://projecteuler.net/problem=35)

The number, $197$, is called a circular prime because all rotations of the digits: $197$, $971$, and $719$, are themselves prime.

There are thirteen such primes below $100$: $2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79$, and $97$.

How many circular primes are there below one million?

In [38]:
# Problem 35 workspace
import itertools as it
def get_digits(n: int):
    return list(map(int, str(n)))



In [43]:
list(it.permutations([1, 2, 3], 3))

[(1, 2, 3), (1, 3, 2), (2, 1, 3), (2, 3, 1), (3, 1, 2), (3, 2, 1)]

In [48]:
def getCircle(n):
    digs = get_digits(n)
    N = n
    cycle = []
    for i in range(len(digs)):
        ones = N%10
        N = N//10
        N += ones*10**(len(digs)-1)
        cycle.append(N)
    return cycle
    

getCircle(1978)

[8197, 7819, 9781, 1978]

In [47]:
from primePy import primes

In [55]:
circular = set()
for n in range(100, int(1e6)):
    if n in circular:
        continue
    if primes.check(n):
        circle = getCircle(n)
        all_prime = True
        for c in circle:
            if not primes.check(c):
                all_prime = False
                break
        if all_prime:
            for c in circle:
                circular.add(c)

len(circular)+13
            
        

55

In [56]:
circular

{113,
 131,
 197,
 199,
 311,
 337,
 373,
 719,
 733,
 919,
 971,
 991,
 1193,
 1931,
 3119,
 3779,
 7793,
 7937,
 9311,
 9377,
 11939,
 19391,
 19937,
 37199,
 39119,
 71993,
 91193,
 93719,
 93911,
 99371,
 193939,
 199933,
 319993,
 331999,
 391939,
 393919,
 919393,
 933199,
 939193,
 939391,
 993319,
 999331}

## Answer: 55

___

# Problem 36
 [Source](https://projecteuler.net/problem=36)

The decimal number, $585 = 1001001001\_2$ (binary), is palindromic in both bases.

Find the sum of all numbers, less than one million, which are palindromic in base $10$ and base $2$.

(Please note that the palindromic number, in either base, may not include leading zeros.)

In [61]:
"abc"[::-1]


0b1001

9

In [69]:
is_palindrome(bin(9)[2:])

True

In [67]:
str(bin(9))

'0b1001'

In [70]:
# Problem 36 workspace

def is_palindrome(n):
    s = str(n)
    return s == s[::-1]

palsum = 0
for i in range(1, int(1e6)):
    if is_palindrome(i) and is_palindrome(bin(i)[2:]):
        palsum += i


palsum

872187

## Answer: 

___

# Problem 37
 [Source](https://projecteuler.net/problem=37)

The number $3797$ has an interesting property. Being prime itself, it is possible to continuously remove digits from left to right, and remain prime at each stage: $3797$, $797$, $97$, and $7$. Similarly we can work from right to left: $3797$, $379$, $37$, and $3$.

Find the sum of the only eleven primes that are both truncatable from left to right and right to left.

NOTE: $2$, $3$, $5$, and $7$ are not considered to be truncatable primes.

In [90]:
import numpy as np
import math

In [96]:
# Problem 37 workspace

math.log10(482345)

5.683357780976457

In [111]:
%timeit 482345 % 10**math.floor(math.log10(482345))

147 ns ± 4.66 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [112]:
%timeit 482345 % 10**int(math.log10(482345)) 

144 ns ± 2.78 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [99]:
math.log10(100000)

5.0

In [ ]:
#
left_trunc(3797)

797

In [126]:
#3797 -> 797
def left_trunc(p):
    return p % 10**int(math.log10(p)) 

def is_left_trunc(p):
    if p == 1:
        return False
    if p<10:
        return primes.check(p)
    else:
        return primes.check(p) and is_left_trunc(left_trunc(p))


is_left_trunc(3797)


True

In [117]:
is_left_trunc(53)

True

In [125]:
def right_trunc(p):
    return p//10

def is_right_trunc(p):
    if p == 1:
        return False
    if p<10:
        return primes.check(p)
    else:
        return primes.check(p) and is_right_trunc(right_trunc(p))
    
is_right_trunc(3797)

True

In [123]:
def is_left_right_trunc(p):
    return is_right_trunc(p) and is_left_trunc(p)

is_left_right_trunc(3797)

True

In [129]:
truncs = set()

n = 11
while len(truncs) < 11:
    if is_left_right_trunc(n):
        truncs.add(n)
        print("Added ", n)
    n += 1

print(sum(truncs))
truncs


Added  23
Added  37
Added  53
Added  73
Added  313
Added  317
Added  373
Added  797
Added  3137
Added  3797
Added  739397
748317


{23, 37, 53, 73, 313, 317, 373, 797, 3137, 3797, 739397}

## Answer: 748317

___

# Problem 38
 [Source](https://projecteuler.net/problem=38)

Take the number $192$ and multiply it by each of $1$, $2$, and $3$:

$$\begin{align}
192 \times 1 &= 192\\
192 \times 2 &= 384\\
192 \times 3 &= 576
\end{align}$$

By concatenating each product we get the $1$ to $9$ pandigital, $192384576$. We will call $192384576$ the concatenated product of $192$ and $(1,2,3)$.

The same can be achieved by starting with $9$ and multiplying by $1$, $2$, $3$, $4$, and $5$, giving the pandigital, $918273645$, which is the concatenated product of $9$ and $(1,2,3,4,5)$.

What is the largest $1$ to $9$ pandigital $9$-digit number that can be formed as the concatenated product of an integer with $(1,2, \dots, n)$ where $n \gt 1$?

In [165]:
# Problem 38 workspace

all_digs = set([n for n in range(1,10)])
best_pandig = 0
best_i = 0

for n in range(2, 6):
    for i in range(1,10000):
        pandig = ''
        for j in range(1,n):
            pandig += str(i*j)
        if len(pandig) == 9 and int(pandig) > best_pandig and set(map(int, pandig)) == all_digs:
            best_i = i
            best_pandig = int(pandig)

best_pandig, best_i

(932718654, 9327)

In [160]:
set(map(int, pandig))

{1, 2, 3, 6, 7, 8, 9}

In [161]:
all_digs

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}

## Answer: 

___

# Problem 39
 [Source](https://projecteuler.net/problem=39)

If $p$ is the perimeter of a right angle triangle with integral length sides, $\{a, b, c\}$, there are exactly three solutions for $p = 120$.

$\{20,48,52\}$, $\{24,45,51\}$, $\{30,40,50\}$

For which value of $p \le 1000$, is the number of solutions maximised?

In [139]:
# Problem 39 workspace

best_solutions = 0
best_p = 0
for p in range(840, 841):
# for p in range(120,121):
    if p%10==0:
        print(p//10, "pct")
    num_solutions = 0

    for a in range(1, p//2):
        for b in range(a, p-2):
            c = p-(a+b)
            if c**2 != a**2 + b**2:
                continue
            if c == int(c):
                print(a, b, c)
                num_solutions += 1
    if num_solutions > best_solutions:
        best_solutions = num_solutions
        best_p = p


84 pct
40 399 401
56 390 394
105 360 375
120 350 370
140 336 364
168 315 357
210 280 350
240 252 348


In [138]:
best_solutions, best_p

(8, 840)

## Answer: 

___

# Problem 40
 [Source](https://projecteuler.net/problem=40)

An irrational decimal fraction is created by concatenating the positive integers:
$$0.12345678910{\color{red}\mathbf 1}112131415161718192021\cdots$$

It can be seen that the $12$
th
digit of the fractional part is $1$.

If $d\_n$ represents the $n$
th
digit of the fractional part, find the value of the following expression.
$$d\_1 \times d\_{10} \times d\_{100} \times d\_{1000} \times d\_{10000} \times d\_{100000} \times d\_{1000000}$$

In [ ]:
# Problem 40 workspace

## Answer: 

___